src/devices/SoftwareSpi/SoftwareSpi.cs

```diff
@@ -0,0 +1,229 @@
+                    _controller.Write(_cs, !(bool)_settings.ChipSelectLineActiveState);
+                });
+        }
+
+        /// <inheritdoc />
+        public override SpiConnectionSettings ConnectionSettings => _settings;
+
+        /// <inheritdoc />
+        public override void TransferFullDuplex(ReadOnlySpan<byte> dataToWrite, Span<byte> dataToRead)
+        {
+            if (dataToRead.Length != dataToRead.Length)
+            {
+                throw new ArgumentException(nameof(dataToWrite));
+            }
+
+            int bitLen = _settings.DataBitLength;
+            int lastBit = bitLen - 1;
+
+            using (StartChipSelect())
+            {
+                for (int i = 0; i < dataToRead.Length; i++)
+                {
+                    byte readByte = 0;
+                    for (int j = 0; j < bitLen; j++)
+                    {
+                        using (StartBitTransfer())
+                        {
+                            int bit = _settings.DataFlow == DataFlow.MsbFirst ? lastBit - j : j;
+                            bool bitToWrite = ((dataToWrite[i] >> bit) & 1) == 1;
+                            if (ReadWriteBit(bitToWrite))
+                            {
+                                readByte |= (byte)(1 << bit);
+                            }
+                        }
+                    }
+
+                    dataToRead[i] = readByte;
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        private bool ReadWriteBit(bool bitToWrite)
+        {
+            _controller.Write(_mosi, bitToWrite);
+            return (bool)_controller.Read(_miso);
+        }
+
+        /// <inheritdoc />
+        public override void Read(Span<byte> data)
+        {
+            Span<byte> dataToWrite = stackalloc byte[data.Length];
+            TransferFullDuplex(dataToWrite, data);
+        }
+
+        /// <inheritdoc />
+        public override void Write(ReadOnlySpan<byte> data)
+        {
+            Span<byte> dataToRead = stackalloc byte[data.Length];
+            TransferFullDuplex(data, dataToRead);
+        }
+
+        /// <inheritdoc />
+        public override void WriteByte(byte data)
+        {
+            Span<byte> outData = stackalloc byte[1];
+            outData[0] = data;
+            Write(outData);
+        }
+
+        /// <inheritdoc />
+        public override byte ReadByte()
+        {
+            Span<byte> data = stackalloc byte[1];
+            Read(data);
+            return data[0];
+        }
+
+        /// <inheritdoc />
+        public override void Dispose(bool disposing)
+        {
+            _controller?.Dispose();
+            _controller = null;
+            base.Dispose();
+        }
+
+        private ScopeData _bitTransfer;
+        private ScopeData _chipSelect;
+        private Scope StartBitTransfer() => new Scope(_bitTransfer);
+        private Scope StartChipSelect() => new Scope(_chipSelect);
+
+        private class ScopeData
```

Comment:
NIT: Would be nice to have in separate file if it stays under device binding.  I guess it would be okay to keep within if moving to core API later.

Reply:
it's currently a nested private class - I don't think our guidelines require moving it (at least I've seen similar patterns elsewhere and noone ever complained)